### DISCUSSION TITLE
OneTime <Value> call

### DISCUSSION THREAD

**Discussion by Muhammadali143904** [arkivanov/Decompose#157] [Q&A] [answered]
Hi, i'm using arkivanov/Decompose for my desktop app, and i have a question, in android i can use MutableSharedFlow/LiveData instead of Value, and i can observe it, and when i observe i set the lifecyle cycle(create/destroy for example), if i set create/destroy it will as oneTimeLiveData/oneTimeFlow, if start/stop it will viewlifecycle owner, i see in decompose is similar but... lifecycle is access from component and not from compose function(content), so how i can't do it

**Comment by arkivanov** [comment, marked as answer]
Could you please share some code, how do you observe in Android with create/destroy and start/stop?

**Reply by Muhammadali143904** [reply]
I do it in xml, i can't do it in compose
So what i must do, for example to show error message which came from server-backend

**Reply by arkivanov** [reply]
You mentioned the following - "in android i can use MutableSharedFlow/LiveData instead of Value, and i can observe it, and when i observe i set the lifecyle cycle(create/destroy for example), if i set create/destroy it will as oneTimeLiveData/oneTimeFlow, if start/stop it will viewlifecycle owner". Could you please show an example of how are you doing that?

**Reply by Muhammadali143904** [reply]
For example in Flow:
`
class ViewModel : ViewModel() {
    // Normal
    var textFieldFlow = MutableSharedFlow(replay = 1, onBufferOverflow = BufferOverflow.DROP_LATEST)
    // One Time
    var messageFlow = MutableSharedFlow(replay = 1, onBufferOverflow = BufferOverflow.DROP_LATEST)

    // Normal
    var textFieldLiveData = MutableLiveData()
    // One Time
    var messageLiveData = MutableLiveData()

    // There are similar, are differents are in view
}
`

And it is view:
`
class Fragment /*For example*/ : Fragment() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // I observe to create/destroy, it as one time
        viewModel.messageFlow
            .onEach { message ->
                view?.let {
                    Snackbar.make(it, message, 2000).show()
                }
            }
            .launchIn(lifecycleScope)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        // I observe it in createView/destroyView
        viewModel.textFieldFlow
            .onEach {
                viewBinding.textField.text = it
            }.launchIn(lifecycleScope)

        // And for liveData
        viewModel.textFieldLiveData.observe(owner = viewLifecycleOwner) {
            viewBinding.textField.text = it
        }
        // One Time
        viewModel.messageLiveData.observe(owner = this@Fragment) {
            Snackbar.make(view, message, 2000).show()
        }
    }
}
`

In Flow main differents in place where we observe(onEach/collect), and in LiveData where we set owner, in the Decompose we can call observe method to Value with similar functionality, but we need a lifecycle which is only in component, and we need observe it in composable function(ui).

**Reply by arkivanov** [reply]
Thanks for samples!

First of all, it is advised to observe events in a component, and just update the state. Ideally, the UI should just render the state of the component, and send events back to the component.

But if you really want to observe a `Flow` or an `Observable` from Compose, then here is an example:

```kotlin

interface Main {
    val events: Flow<Event>

    sealed interface Event {
        // ...
    }
}

@Composable
fun MainContent(main: Main) {
    LaunchedEffect(main) {
        main.events.collect { event ->
            // ...
        }
    }
}
```

**Reply by Muhammadali143904** [reply]
Ok, thank you arkivanov, and last sorry for question, how you will show the message wich came from server-backend, use in compose and your/Decompose, and thank you again

**Reply by arkivanov** [reply]
Sure no problem! With coroutines you can do it as follows:

```kotlin
interface Main {
    val events: Flow<Event>

    sealed interface Event {
        data class Error(val message: String) : Event
    }
}

@Composable
fun MainContent(main: Main) {
    val snackbarHostState = remember { SnackbarHostState() }

    Scaffold(
        modifier = Modifier.fillMaxSize(),
        scaffoldState = rememberScaffoldState(snackbarHostState = snackbarHostState),
    ) {
        Text(text = "Hello")
    }

    LaunchedEffect(main) {
        main.events.collect {
            when (it) {
                is Main.Event.Error -> snackbarHostState.showSnackbar(it.message)
            }
        }
    }
}
```

**Reply by Muhammadali143904** [reply]
Thank you much much much much, and sorry for my stupidity =)

**Reply by arkivanov** [reply]
There are no stupid questions, feel free to ask!

**Reply by Muhammadali143904** [reply]
Ok, thank you again...
